server/protocols/jmap/src/main/java/org/apache/james/jmap/model/MessageContentExtractor.java

```diff
@@ -90,18 +94,41 @@ private String asString(TextBody textBody) throws IOException {
     }
 
     private MessageContent retrieveHtmlAndPlainTextContent(Multipart multipart) throws IOException {
-        Optional<String> textBody = getFirstMatchingTextBody(multipart, "text/plain");
-        Optional<String> htmlBody = getFirstMatchingTextBody(multipart, "text/html");
-        return new MessageContent(textBody, htmlBody);
+        Optional<String> textBody = getFirstMatchingTextBody(multipart, TEXT_PLAIN);
+        Optional<String> htmlBody = getFirstMatchingTextBody(multipart, TEXT_HTML);
+        MessageContent directChildTextBodies = new MessageContent(textBody, htmlBody);
+        if (!directChildTextBodies.isComplete()) {
+            MessageContent fromInnerMultipart = parseFirstFoundMultipart(multipart);
+            return directChildTextBodies.merge(fromInnerMultipart);
+        }
+        return directChildTextBodies;
+    }
+
+    private MessageContent retrieveFirstReadablePart(Multipart multipart) throws IOException {
+        return multipart.getBodyParts()
+            .stream()
+            .filter(this::isNotAttachment)
+            .flatMap(Throwing.function(this::returnIfReadable).sneakyThrow())
+            .findFirst()
+            .orElse(MessageContent.empty());
     }
 
-    private MessageContent retrieveFirstHtmlOrPlainTextContent(Multipart multipart) throws IOException {
-        Optional<String> textBody = Optional.empty();
-        Optional<String> htmlBody = getFirstMatchingTextBody(multipart, "text/html");
-        if (! htmlBody.isPresent()) {
-            textBody = getFirstMatchingTextBody(multipart, "text/plain");
+    private Stream<MessageContent> returnIfReadable(Entity entity) throws IOException {
```

Comment:
this method has no notion of "first" as the findFirst is done later. Changed to extractContentIfReadable